src/adaptix/_internal/morphing/generic_provider.py

```diff
@@ -410,14 +415,54 @@ def _provide_dumper(self, mediator: Mediator, request: DumperRequest) -> Dumper:
         dumper_type_dispatcher = ClassDispatcher(
             {type(None) if case.origin is None else case.origin: dumper for case, dumper in zip(norm.args, dumpers)},
         )
-        return self._get_dumper(dumper_type_dispatcher)
 
-    def _get_dumper(self, dumper_type_dispatcher: ClassDispatcher[Any, Dumper]) -> Dumper:
+        literal_dumper = self._get_dumper_for_literal(norm, dumpers, dumper_type_dispatcher)
+
+        if literal_dumper:
+            return literal_dumper
+
+        return self._produce_dumper(dumper_type_dispatcher)
+
+    def _produce_dumper(self, dumper_type_dispatcher: ClassDispatcher[Any, Dumper]) -> Dumper:
         def union_dumper(data):
             return dumper_type_dispatcher.dispatch(type(data))(data)
 
         return union_dumper
 
+    def _produce_dumper_for_literal(
+        self,
+        dumper_type_dispatcher: ClassDispatcher[Any, Dumper],
+        literal_dumper: Dumper,
+        literal_cases: Sequence[Any],
+    ) -> Dumper:
+        def union_dumper_with_literal(data):
+            if data in literal_cases:
+                return literal_dumper(data)
+            return dumper_type_dispatcher.dispatch(type(data))(data)
+
+        return union_dumper_with_literal
+
+    def _get_dumper_for_literal(
+        self,
+        norm: BaseNormType,
+        dumpers: Iterable[Any],
+        dumper_type_dispatcher: ClassDispatcher[Any, Dumper],
+    ) -> Optional[Dumper]:
+        try:
+            literal_type, literal_dumper = next(
+                (union_case, dumper) for union_case, dumper
+                in zip(norm.args, dumpers)
+                if union_case.origin is Literal
+            )
+        except StopIteration:
+            literal_type, literal_dumper = None, None
+
+        if not literal_type:
+            return None
```

Comment:
You can return from StopIteration exception handler

Reply:
done